src/test/codewhisperer/service/telemetry.test.ts

```diff
@@ -0,0 +1,656 @@
+            ])
+        })
+
+        it('reject: trigger then close editor', async function () {
+            assertSessionClean()
+            const editor = await openATextEditorWithText('', 'test.py')
+
+            await manualTrigger(editor, client, config)
+            await waitUntilSuggestionSeen()
+            await closeActiveEditor()
+            // assert.strictEqual(editor.document.getText(), '')
+
+            assertTelemetry('codewhisperer_userTriggerDecision', [
+                session1UserTriggerEvent({ codewhispererSuggestionState: 'Reject' }),
+            ])
+        })
+
+        it('reject: onFocusChange', async function () {
+            assertSessionClean()
+            const editor = await openATextEditorWithText('', 'test.py')
+
+            await manualTrigger(editor, client, config)
+            await waitUntilSuggestionSeen()
+            // assert.strictEqual(editor.document.getText(), '')
+
+            await RecommendationHandler.instance.onFocusChange()
+            assertTelemetry('codewhisperer_userTriggerDecision', [
+                session1UserTriggerEvent({ codewhispererSuggestionState: 'Reject' }),
+            ])
+        })
+    })
+})
+
+async function manualTrigger(
+    editor: vscode.TextEditor,
+    client: DefaultCodeWhispererClient,
+    config: ConfigurationEntry
+) {
+    await invokeRecommendation(editor, client, config)
+}
+
+// Note: RecommendationHandler.isSuggestionVisible seems not to work well, hence not using it
+async function waitUntilSuggestionSeen(index: number = 0) {
+    const state = await waitUntil(
+        async () => {
+            const r = session.getSuggestionState(index)
+            if (r) {
+                return r
+            }
+        },
+        {
+            interval: 50,
+        }
+    )
+
+    assert.ok(state === 'Showed')
+    await sleep(50)
+}
+
+async function acceptByTab() {
+    await vscode.commands.executeCommand('editor.action.inlineSuggest.commit')
+
+    // required because oninlineAcceptance has sleep(vsCodeCursorUpdateDelay), otherwise assertion will be executed before onAcceptance hook
+    await sleep(vsCodeCursorUpdateDelay + 50)
+}
+
+async function rejectByEsc() {
+    return vscode.commands.executeCommand('aws.codeWhisperer.rejectCodeSuggestion')
+}
+
+async function navigateNext() {
+    return vscode.commands.executeCommand('editor.action.inlineSuggest.showNext')
+}
+
+async function navigatePrev() {
+    return vscode.commands.executeCommand('editor.action.inlineSuggest.showPrevious')
+}
+
+async function closeActiveEditor() {
+    return vscode.commands.executeCommand('workbench.action.closeActiveEditor')
+}
+
+async function typing(editor: vscode.TextEditor, s: string) {
+    for (const char of s) {
+        await typeAChar(editor, char)
+    }
+    // waitUntilSuggestionSeen doesn't work for this
+    // see if we can find another way to wait
+    await sleep(100)
```

Comment:
have tried different combination of `waitUntil`, but none of them really worked. If we don't wait some time after typing them directly call `commit` command, there will be no effect because VSCode is still resolving the typeahead and trying to provide the adjusted inline suggestion.

Reply:
waitUntil has an `interval` option to avoid a "hot loop": https://github.com/aws/aws-toolkit-vscode-staging/blob/a19a4aaf350ac18ddda0e1a0998afd9f865b0b2c/src/shared/utilities/timeoutUtils.ts#L177-L178

does that help?

Reply:
the problem here is more like we don't have a exact target to wait until since we don't have an API like `isInlineSuggestionVisible` or something.  And if I call `editor.action.inlineSuggestion.commit` before it's ready, there will be nothing happen..

Reply:
let me try with higher interval value

Reply:
ahh I slightly update the function `acceptByTab()` here https://github.com/aws/aws-toolkit-vscode/pull/4046/files#diff-2dc09778984ac96fa6f8170184bf439825bf4839775915c3c8e5da4d2cc47c25R546

Reply:
`assertEditorContains` works perfectly, the only thing left is the waiting until inline suggestion is shown then accept

Reply:
so move waiting logic inside `acceptByTab()`, but I felt I can't avoid the very first line of `await sleep(x)` in this `waitUntil` otherwise it just won't work as expected because we don't have a `isInlineSuggestionActive` API to waitUntil